[red-knot]: Add a VendoredFileSystem implementation

AlexWaygood opened the pull request:
## Summary

Work towards https://github.com/astral-sh/ruff/issues/11653.

This PR adds a `VendoredFileSystem` implementation to the `ruff_db` crate, the idea being that this is the filesystem we would use for resolving vendored typeshed stubs.

Some of this feels a bit awkward, but I'm not sure there's a better way. I ran into two problems while writing this PR:
1. The `ZipArchive` struct exposed by the `zip` crate takes `&mut self` for most of the methods that query the underlying archive for information about files or directories within the archive. This means that in order to query zipped data from trait method implementations that take `&self`, we end up having to clone the underlying archive. I think that's okay, as the `zip` docs say that it should be cheap to clone (at least right now), but it didn't fill me with joy.
2. The `camino` crate strips trailing slashes in a lot of its methods, as they're not semantically significant for most paths. However, for zip archives, trailing slashes are very significant: if a `stdlib/` directory exists inside the zip, `zip.by_name("stdlib")` will not resolve to that directory (only `zip.by_name("stdlib/")` will). I've attempted to abstract over this difference to other kinds of paths, so that the `VendoredFileSystem` is consistent with the other `FileSystem` implementations in this respect. But it was slightly tricky to do so.

## Test Plan

I added lots of parameterized tests.

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+    }
+}
+
+impl FileSystem for VendoredFileSystem {
```
We shouldn't implement `FileSystem` for vendored paths. I think that's also the part that @carljm felt some confusion around. There are a few reasons:

* `FileSystem` is an abstraction over `std::fs`. It's not a way of implementing an overlay filesystem that supports reading files from different sources. That's that `Vfs` does
* `FileSystem` works on `FileSystemPath`, but we want to use `VendoredFilePath` and `VendoredFilePathBuf` for vendored file. We intentionally use different types to prevent that someone tries to read a `VendoredPath` from the file system or the other way round
* Many `FileSystem` operations are simply irrelevant for `VendoredPath`s. The only thing that we need to support for vendored path is testing if a file exists (or a simplified metadata that returns the last modified date), and reading the content. We won't need file watching, directory traversal, or operations for testing if something's a directory. We only care about the files, and nothing else

That's why I think we can either implement the methods as free-standing functions similar to `fs::read_to_string`, or as a struct, but where we define our own functions.

the vendored path because vendored path should be represented using `VendoredPath` and `VendoredPathBuf`. 

The other reason is that most operations aren't releve

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+    }
+}
+
+/// A path that has been normalized via the `normalize_vendored_path` function.
+///
+/// Trailing slashes are normalized away by `camino::Utf8PathBuf`s,
+/// but trailing slashes are crucial for distinguishing between
+/// files and directories inside zip archives.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct NormalizedVendoredPath(String);
```
I think we could combine this with the `VendoredFilePath` and `VendoredFilePathBuf` where we require the use of `VendoredFilePathBuf::new` (that normalizes the string).

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
I would recommend that we move all the vendored code into its own `ruff_vendored` crate, along with the files and `build.rs`.

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+impl VendoredFileSystem {
+    const READ_ONLY_PERMISSIONS: u32 = 0o444;
+
+    pub fn new(raw_bytes: Arc<[u8]>) -> Result<Self> {
```
Can we use a `&'static [u8]` instead of an `Arc<[u8]>`
```suggestion
    pub fn new(raw_bytes: &'static [u8]) -> Result<Self> {
```

AlexWaygood replied:
I had that initially, but it made it much harder to test it using a mock zipfile

AlexWaygood replied:
This relates to the conversation we're having at https://github.com/astral-sh/ruff/pull/11863#discussion_r1639342264. If we just committed a binary zip to the repo and used that in the test, then we could easily use `&'static` for the implementation. But IMHO that would be a really bad test, because we wouldn't be able to easily see what we'd put in the zip archive, so I feel like any assertions we'd make would honestly not have that much value

MichaReiser replied:
That's fair. I think we should allow both then. Using `&'static [u8]` has the benefit that we can pass the bytes directly from the binary without having to copy the zip to the heap first (it's already on the heap because the binary is on the heap, but we can avoid duplicating it). 

You can implement it by having a 

```rust
enum ZipData {
	Static(&'static [u8]),
	Dynamic(Arc<[u8]) // Or Box if we can drop the `Clone` requirement
}
```

And then provide a  `new_static` and `new` function on the `VendoredFilesystem`

AlexWaygood replied:
Brilliant, thank you!

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+        // Zip lookups require a mutable reference to the archive,
+        // which creates a nightmare with lifetimes unless you own the archive, hence the clone.
+        //
+        // The zip documentation states that:
+        //
+        //     At the moment, this type is cheap to clone if this is the case for the reader it uses.
+        //     However, this is not guaranteed by this crate and it may change in the future.
+        //
+        // So this should be fairly cheap *for now*.
+        self.inner.clone()
```
I'm not sure if I want to rely on this behavior. It feels very easy to get wrong when updating the crate. 

I think what I would do is to wrap the `ZipArchive` in a mutex instead and require explicitly locking. This will reduce concurrency a bit but I don't expect that any of the zip operations take long.

AlexWaygood replied:
This is the problem I described as point (1) in my PR description. Because `ZipArchive::by_name` takes `&mut self`, it's very hard to implement `VendoredFileSystem::read()` (which must take `&self`). I think our options are as follows:
1. Creating a new `ZipArchive` instance every time `read()` is called
2. Using the existing `ZipArchive` instance every time `read()` is called, but cloning it to create an owned copy
3. Accept that `VendoredFileSystem::read()` will need to take `&mut self` rather than `&self` (I don't like this at all).
4. Unzip typeshed eagerly at runtime so that we don't need to use the `zip` crate at all to read vendored files.

Currently I'm doing Option (2) in this PR, but maybe I'm missing something.

AlexWaygood replied:
I switched the PR to use Option (6): "Something something interior mutability"

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+        self.inner.clone()
+    }
+
+    /// Create a new empty string, read the contents of `zipfile` into it, and return the string
+    fn read_zipfile(mut zipfile: ZipFile) -> Result<String> {
```
Nit: This comment explains the implementation almost 1:1. I prefer comments that are either higher level or to omit the comment entirely, because it doesn't provide any additional information and is only at risk of being outdated.

MichaReiser replied:
Nit
```suggestion
    fn read_zipfile(mut zip_file: ZipFile) -> Result<String> {
```

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+        if normalized.has_trailing_slash() {
+            return Err(lookup_error);
+        }
+        let mut archive = self.vendored_archive();
```
Can't we re-use `archive` here?
```suggestion
```

AlexWaygood replied:
No, or the borrow checker complains that we're mutably borrowing it twice in the same scope.

MichaReiser replied:
This seems to compile fine

```rust
let normalized = normalize_vendored_path(path);
        let mut archive = self.vendored_archive();
        let lookup_error = match archive.lookup_path(&normalized) {
            Ok(zipfile) => return Self::read_zipfile(zipfile),
            Err(err) => err,
        };
        if normalized.has_trailing_slash() {
            return Err(lookup_error);
        }
        // let mut archive = self.vendored_archive();
        let zipfile = archive.lookup_path(&normalized.with_trailing_slash())?;
        Self::read_zipfile(zipfile)
```

AlexWaygood replied:
So it does. I'm sure I tried that yesterday. Ah well -- thanks!

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+            Ok(zipfile) => return Self::read_zipfile(zipfile),
+            Err(err) => err,
+        };
+        if normalized.has_trailing_slash() {
```
What's the retry logic here? I get the impression that we do some trial and error here to read the file and that feels off. Can't we just return an error and ask callers to construct the right path instead of guessing paths?

AlexWaygood replied:
This is point (2) that I described in the PR description. For other file systems we use in Ruff, `file_system.exists("stdlib")` will return `true` if there exists a directory called `stdlib`. But without this retry logic, that will return `false` for the vendored filesystem, because zip archives use trailing slashes internally to distinguish between files and directories in the archive. So if there's a `stdlib` directory, then (without this retry logic) `file_system.exists("stdlib/")` would return `true`, but `file_system.exists("stdlib")` would return `false`.

I wanted to try to abstract over that difference in my implementation here, as I think it would be a bug magnet for paths to work differently for the vendored file system compared to how they work with other file systems we're using elsewhere in Ruff.

AlexWaygood replied:
...but this doesn't make any sense, because you can't read a directory, you can only read a file 🤦‍♂️

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+
+    /// Create a new empty string, read the contents of `zipfile` into it, and return the string
+    fn read_zipfile(mut zipfile: ZipFile) -> Result<String> {
+        let mut buffer = String::with_capacity(zipfile.size() as usize);
+        zipfile.read_to_string(&mut buffer)?;
+        Ok(buffer)
```
I think we can just use `read_to_string` here. I haven't followed all the code but `zipfile.read_to_string` calls into `std::read_to_string` and I'm pretty sure they correctly resize the buffer before writing into it.

AlexWaygood replied:
You have to pass in a pre-existing buffer for the `read_to_string` method

MichaReiser replied:
Sorry. What I meant is we can simplify this to 

```rust
let mut buffer = String::new();
zipfile.read_to_string(&buffer)?;
Ok(buffer)
```

Because `read_to_string` already handles the `with_capacity` (it should call `reserve` or similar)

AlexWaygood replied:
Got it, thanks!

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+        };
+
+        Ok(Metadata {
+            revision: FileRevision::zero(),
```
We should use a revision other than zero or we might run into problems when the vendored file system content changes. We could use ruff's version, the last modified time of the files in the zip file, or the time when the `VendoredFileSystem` was created.

AlexWaygood replied:
I thought one of the guarantees of the vendored file system was that the files could never change within a single invocation of Ruff. Or is the concern here that this would lead to incorrect data being loaded from the persistent on-disk cache on a subsequent run of Ruff?

MichaReiser replied:
It should probably be fine without it, because we should bust the entire cache if the ruff version changes. But having it in the revision gives us an extra layer of protection, because salsa would then invalidate the queries for us, in case we forget.

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+        components.next();
+    }
+
+    for component in components {
+        match component {
+            camino::Utf8Component::Normal(part) => normalized_parts.push(part),
+            camino::Utf8Component::CurDir => continue,
+            camino::Utf8Component::ParentDir => {
+                normalized_parts.pop();
+            }
+            unsupported => panic!("Unsupported component in a vendored path: {unsupported}"),
+        }
+    }
```
We have almost the same logic in `memory` rs 😆

AlexWaygood replied:
Not a coincidence 😉 I looked at that function quite a bit while working on this, but in the end I concluded that there was enough that needed to be different that it would require a separate function.

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+    const ASYNCIO_TASKS_CONTENTS: &str = "class Task: ...";
+
+    static MOCK_ZIP_ARCHIVE: Lazy<Arc<[u8]>> = Lazy::new(|| {
+        let td = TempDir::new().unwrap();
+        let typeshed_path = td.path().join("typeshed");
+        let typeshed = fs::File::create(&typeshed_path).unwrap();
```
I don't think we need a file here. You can use a `Vec<u8>` as a target and just write into that. 

But what could be simpler (and closer to the actual code) is to create a zip file manually  and commit it. You can then use `include_bytes!` to include the bytes. 

This also avoids creating a zip file on every test run.

AlexWaygood replied:
> But what could be simpler (and closer to the actual code) is to create a zip file manually and commit it. You can then use `include_bytes!` to include the bytes.
> 
> This also avoids creating a zip file on every test run.

I considered this. But I thought it made for a much more readable test if you could see exactly which files are being put in the mocked zip as part of the test setup. Otherwise it feels unclear what exactly we're testing.

MichaReiser replied:
That makes sense. But we should replace `tempfile` with a `&mut Vec<u8>`

AlexWaygood replied:
Yeah, I'll make that change for sure!

MichaReiser commented on crates/ruff_db/src/file_system/vendored.rs:
```diff
@@ -0,0 +1,353 @@
+        VendoredFileSystem::new(Arc::clone(&MOCK_ZIP_ARCHIVE)).unwrap()
+    }
+
+    macro_rules! directory_tests {
+        ($($name:ident: $dirname:literal,)*) => {
+        $(
+            #[test]
+            fn $name() {
+                let mock_typeshed = mock_typeshed();
+
+                let path = FileSystemPath::new($dirname);
+
+                assert!(mock_typeshed.exists(path));
+                assert!(mock_typeshed.is_directory(path));
+                assert!(!mock_typeshed.is_file(path));
+
+                let path_metadata = mock_typeshed.metadata(path).unwrap();
+                assert!(path_metadata.file_type().is_directory());
+                assert!(path_metadata.permissions().is_some());
+                assert_eq!(path_metadata.revision(), FileRevision::zero());
+            }
+        )*
+        }
+    }
+
+    directory_tests! {
+        stdlib_dir_no_trailing_slash: "stdlib",
+        stdlib_dir_trailing_slash: "stdlib/",
+        asyncio_dir_no_trailing_slash: "stdlib/asyncio",
+        asyncio_dir_trailing_slash: "stdlib/asyncio/",
+        stdlib_dir_parent_components: "stdlib/asyncio/../../stdlib",
+        asyncio_dir_odd_components: "./stdlib/asyncio/../asyncio/",
+    }
```
I think we don't need to use macros here. They safe very little code but break IDE features. For example, I can't jump or run the test manually anymore. Rustfmt breaks down, and auto completion often doesn't work. They're also hard to read. 

```suggestion
    fn assert_directory(directory_name: &str) {
                let mock_typeshed = mock_typeshed();

                let path = FileSystemPath::new(directory_name);

                assert!(mock_typeshed.exists(path));
                assert!(mock_typeshed.is_directory(path));
                assert!(!mock_typeshed.is_file(path));

                let path_metadata = mock_typeshed.metadata(path).unwrap();
                assert!(path_metadata.file_type().is_directory());
                assert!(path_metadata.permissions().is_some());
                assert_eq!(path_metadata.revision(), FileRevision::zero());
    }
    
    #[test]
    fn stdlib_dir_no_trailing_slash() {
      assert_directory("stdlib");
    }
```

Same for the other tests

AlexWaygood replied:
Duh. I felt so clever about this... How did I forget about this much simpler solution 😑

(though FWIW, it didn't seem to break any IDE features for me on VSCode — I was still able to run the test manually locally)

MichaReiser reviewed (changes requested):
Nice work. 

Overall looks good, but we should address the following items before merging:

* Use regular functions for tests instead of macros
* Remove `FileSystem` implementation and use `VendoredFilePath` and `VendoredFilePathBuf` (and move the definitions into `vendored.rs)
* Move the implementation out of the `vendored` directory.
* If we can, remove the probing where we append a slash at the end of the path when reading failed. This seems like trial and error. 

One open question for me is whether we should create a dedicated `ruff_vendored` crate that also contains all vendored files (and contains the `build.rs`). The benefit of that would be that integration tests can use the actual vendored files, which I think would be desired (but use stubed vendored files for unit tests). If that's undesired, I think leaving it in `ruff_db` is fine.

AlexWaygood commented:
> * Move the implementation out of the `vendored` directory.

I'm guessing you mean `file_system` directory? :)

AlexWaygood commented:
> One open question for me is whether we should create a dedicated `ruff_vendored` crate that also contains all vendored files (and contains the `build.rs`). The benefit of that would be that integration tests can use the actual vendored files, which I think would be desired (but use stubed vendored files for unit tests). If that's undesired, I think leaving it in `ruff_db` is fine.

I wasn't sure while writing this the extent to which we'd want to tie the `VendoredFileSystem` implementation to the exact details of the only repository we realistically expect to be vendoring (typeshed). In this implementation, I tried to keep things relatively abstract, so that it could theoretically work with any vendored zip archive. That means that I expect to have to build another layer of abstraction on top of this for working with typeshed specifically, because we have to take the `VERSIONS` file into account when deciding whether a file "exists" in typeshed. If a file `stdlib/foo.pyi` exists in the typeshed zip archive but `VERSIONS` tells us that it hadn't actually been added to the stdlib yet on the Python version the user has selected to be the target version, then the query to our typeshed abstraction should arguably report that the `stdlib/foo.pyi` doesn't exist at all. That's the reality at runtime on that Python version, after all: if the module hasn't been added to the stdlib yet, the file for that module physically doesn't exist at runtime on that Python version.

MichaReiser commented on crates/ruff_db/src/file_system.rs:
```diff
@@ -7,9 +7,11 @@ use filetime::FileTime;
 
 pub use memory::MemoryFileSystem;
 pub use os::OsFileSystem;
+pub use vendored::VendoredFileSystem;
```
I suggest moving `vendored` into its own module, considering that it isn't implementing `FileSystem`.

AlexWaygood replied:
Yes, I'm going to make that change. It still implements `FileSystem` in this PR currently, but I'm working on changing that.

AlexWaygood commented:
@MichaReiser -- I think I've addressed all your review comments now except for https://github.com/astral-sh/ruff/pull/11863#discussion_r1639324437 (where we want to do the path normalization). As I mentioned on our call earlier, I think I'd prefer to experiment with that in a followup, if that's okay, and see if we think it improves the API or not. So I think this PR should be ready for re-review now!

AlexWaygood commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,379 @@
+        static DATA: Lazy<Box<[u8]>> = Lazy::new(|| {
+            let mut buffer = Vec::new();
+            let cursor = io::Cursor::new(&mut buffer);
+            {
+                let mut archive = ZipWriter::new(cursor);
+                archive.finish().unwrap();
+            }
```
Style question: do you prefer an isolated scope like I currently have, or a call to `drop()`, for this kind of pattern? Not sure which is considered more idiomatic.

```suggestion
            let mut archive = ZipWriter::new(cursor);
            archive.finish().unwrap();
            drop(archive);
```

MichaReiser replied:
I prefer the sub scopes.

MichaReiser commented on crates/ruff_db/src/vfs.rs:
```diff
@@ -315,7 +324,7 @@ impl VendoredVfs {
     fn read(&self, path: &VendoredPath) -> Option<String> {
         match self {
-            VendoredVfs::Real => todo!(),
+            VendoredVfs::Real(file_system) => file_system.read(path),
```
Nit: Maybe rename to `vendored` to avoid confusion with `FileSystem`
```suggestion
            VendoredVfs::Real(vendored) => vendored.read(path),
```

AlexWaygood replied:
It _is_ a file system though, it just isn't an _OS_ file system :) that's what the `fs` in `Vfs` stands for, right?

I don't think `vendored` as a name works well here, because the distinction between the two enum branches is whether it's a "Real file system" or a "Stubbed file system", not whether it's a "Real vendored" or a "Stubbed vendored" (that doesn't really make sense).

I would prefer it if we considered renaming the `FileSystem` trait instead

MichaReiser replied:
Yeah not sure. The `FileSystem` maps to `std::fs` and vendored is not that. I don't think `FileSystem` maps to the `OS` file system, because an in memory (or wasm FS) isn't the OS's file system.

I'm currently thinking about renaming `VfsFile` to `Files`, `VfsFile` to `File` (not sure about `VfsPath`). `VendoredFiles` would then fit well into this naming and would also make it clear, that it only supports files. But I'm okay deferring this for now.

MichaReiser commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,379 @@
+type Result<T> = io::Result<T>;
+
+#[derive(Debug)]
+pub struct VendoredFileSystem {
```
Nit: I would just call this `Vendored` or `VendoredFiles` to avoid any confusion with `FileSystem`.

MichaReiser commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,379 @@
+        let mut archive = inner_locked.borrow_mut();
+        let mut zip_file = archive.lookup_path(&normalize_vendored_path(path)).ok()?;
+        let mut buffer = String::new();
+        zip_file.read_to_string(&mut buffer).ok()?;
```
What are the reasons why `read_to_string` can fail? I'm not sure if just silently returning `None` in that case is the best solution. Because the file does exist, it just can't be read for obscure reasons. We may want to change this method to return a result instead.

AlexWaygood replied:
It returns an `Err` if the data that it is trying to read into the `String` buffer turns out not to be valid UTF-8. I agree that we shouldn't silently return `None` here if that happens.

MichaReiser commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,379 @@
+struct VendoredZipArchive(ZipArchive<io::Cursor<&'static [u8]>>);
+
+impl VendoredZipArchive {
+    pub fn new(data: &'static [u8]) -> Result<Self> {
```
```suggestion
    fn new(data: &'static [u8]) -> Result<Self> {
```

MichaReiser commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,379 @@
+    inner: VendoredFileSystemInner,
+}
+
+impl Default for VendoredFileSystem {
+    fn default() -> Self {
+        static DATA: Lazy<Box<[u8]>> = Lazy::new(|| {
+            let mut buffer = Vec::new();
+            let cursor = io::Cursor::new(&mut buffer);
+            {
+                let mut archive = ZipWriter::new(cursor);
+                archive.finish().unwrap();
+            }
```
What's the use case for implementing `Default`? I think  I strongly prefer not to implement `Default` for this type because it just isn't very useful when empty. 

If you want to support an empty `VendoredFS`, then I think I would model it as an enum with `EMPTY` and `Data(&'static [u8]`) variants.

AlexWaygood replied:
I implemented `Default` here because you'd implemented `Default` on the `VendoredVfs` enum, and I assumed you had a reason for that 😆

I don't see a great reason for implementing `Default` on this struct; I can get rid of it.

MichaReiser commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,379 @@
+/// Unsupported components are path prefixes,
+/// and path root directories appearing anywhere except at the start of the path.
+fn normalize_vendored_path(path: &VendoredPath) -> NormalizedVendoredPath {
+    debug_assert!(
+        !path.as_std_path().is_absolute(),
+        "Attempted to lookup an absolute filesystem path relative to a vendored zip archive"
+    );
+    let mut normalized_parts = camino::Utf8PathBuf::new();
```
We don't want to use `is_absolute`. It's platform dependent, meaning `VendoredPath`'s suddenly become platform specific. 

If you want to use linux semantics, use `!path.starts_with("/")`

AlexWaygood replied:
It's very surprising to me that `is_absolute` is platform specific. Thanks for the reminder!

AlexWaygood replied:
Oh wait, no, you said "platform-dependent", not "platforms-specific". I think "platform-dependent" is what I want here. Whatever kind of path we're dealing with here, it should never have been resolved to an absolute OS-filesystem path. It should always be a path that's relative to the root of the zip directory. If this assertion ever fails on either Linux or Windows, I think that indicates a logical error in our code somewhere.

MichaReiser replied:
A path starting with `/` will fail on Windows.

AlexWaygood replied:
I'm still not sure I fully understand. The main purpose of this debug assertion is to make sure that the path hasn't been `canonicalize`d or otherwise "made absolute" prior to attempting to resolve the path relative to the zip archive. On a Unix system `canonicalize()` would mean that the path ould start with `/`, but on a Windows system it would mean that it would start with a prefix, right? So I think `is_absolute()` is what we want here?

MichaReiser replied:
I think we want `VendoredPath` to be file system agnostic. It's not a real file system so we should avoid the madness of having different path representations for different systems. I'm also not sure what happens if you lookup the path `C:\test` in the zip file. I would simply assume that this would outright fail?

I recommend that `VendoredPath` uses unix representation exclusively.

AlexWaygood replied:
`VendoredPath` does use unix representation exclusively; this PR doesn't change that, and nor does this assertion. I think my takeaway here is that this debug assertion causes more confusion than clarification, so I think I'll just remove it altogether :)

MichaReiser replied:
The debug assertion itself is fine, but it can't be `is_absolute` because `is_absolute` returns `false` on Windows if the path starts with `/`

AlexWaygood replied:
I understand that. I still think we're talking past each other a little bit, but I've got rid of the assertion now anyway.

MichaReiser commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,379 @@
+        Ok(Self(ZipArchive::new(io::Cursor::new(data))?))
+    }
+
+    fn lookup_path(&mut self, path: &NormalizedVendoredPath) -> Result<ZipFile> {
+        Ok(self.0.by_name(path.as_str())?)
+    }
```
Should `lookup_path` take a `&VendoredPath` instead and do all the normalization?

AlexWaygood replied:
Because of the "trailing slashes are semantically significant for zip archives" thing combined with the "`ZipFile::by_name()` requires a mutable reference" thing, it's hard for me to see how we could do this while keeping the borrow checker happy and avoiding doing path normalization twice. It doesn't seem possible to call `self.0.by_name()` twice in `lookup_path()` while keeping the borrow checker happy, unfortunately (unless I'm missing something), and because of the trailing-slash thing, we need to call it twice in `lookup_path()` if we're going to do the normalization inside `lookup_path()`.

MichaReiser reviewed (commented):
This overall looks good but I think we can make it much simpler by removing all code that's only necessary for handling directories. From what I understand, we won't need directory support to implement module resolution. All we need is to lookup a specific file path. Or are there operations that need to know about the existence of a directory?

AlexWaygood commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,379 @@
+        })
+    }
+
+    pub fn file_type(&self, path: &VendoredPath) -> Option<FileType> {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        if let Ok(zip_file) = archive.lookup_path(&normalized) {
+            if zip_file.is_dir() {
+                return Some(FileType::Directory);
+            }
+            return Some(FileType::File);
+        }
+        if let Ok(zip_file) = archive.lookup_path(&normalized.with_trailing_slash()) {
+            if zip_file.is_dir() {
+                return Some(FileType::Directory);
+            }
+            return Some(FileType::File);
+        }
+        None
+    }
```
Yes, I wasn't sure if this was necessary or not, so I think you're right that it probably makes sense to remove it for now until there's a need for it.

AlexWaygood commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,379 @@
+            .lookup_path(&normalized.with_trailing_slash())
+            .map(|zip_file| zip_file.crc32())
+            .ok()
+    }
```
(As discussed offline, we probably are going to need more than these two APIs, since you can have subpackages and namespace packages inside the stdlib just like in any other Python directories)

MichaReiser commented on crates/ruff_db/src/metadata.rs:
Maybe rename the module to `file_revision`, considering that it doesn't contain anything else.

MichaReiser commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,351 @@
+                .is_ok()
+    }
+
+    pub fn metadata(&self, path: &VendoredPath) -> Option<Metadata> {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        if let Ok(zip_file) = archive.lookup_path(&normalized) {
+            return Some(Metadata::from_zip_file(zip_file));
+        }
+        if let Ok(zip_file) = archive.lookup_path(&normalized.with_trailing_slash()) {
+            return Some(Metadata::from_zip_file(zip_file));
+        }
+        None
+    }
```
I think some empty lines could help with readability:

* setup
* handle possible file
* handle possible directory

```suggestion
    pub fn metadata(&self, path: &VendoredPath) -> Option<Metadata> {
        let normalized = normalize_vendored_path(path);
        let inner_locked = self.inner.lock();
        let mut archive = inner_locked.borrow_mut();

        if let Ok(zip_file) = archive.lookup_path(&normalized) {
            return Some(Metadata::from_zip_file(zip_file));
        }
        
        if let Ok(zip_file) = archive.lookup_path(&normalized.with_trailing_slash()) {
            return Some(Metadata::from_zip_file(zip_file));
        }
        None
    }
```

MichaReiser commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,351 @@
+type Result<T> = io::Result<T>;
+
+#[derive(Debug)]
+pub struct VendoredFileSystem {
```
It might be helpful to add some documentation to this struct, even if it's very brief

MichaReiser commented on crates/ruff_db/src/vendored.rs:
```diff
@@ -0,0 +1,351 @@
+    }
+
+    pub fn exists(&self, path: &VendoredPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        archive.lookup_path(&normalized).is_ok()
+            || archive
+                .lookup_path(&normalized.with_trailing_slash())
+                .is_ok()
```
I would find some documentation helpful that explains the difference between paths ending with trailing slash and once that don't. Because it isn't clear from reading this code what's happening here and why proving the zip two times is necesary.

MichaReiser replied:
Nit: Maybe just `self.metadata(path).is_some()`
